src/utils/VoronoiUtils.cpp

```diff
@@ -0,0 +1,279 @@
+        break;
+    case boost::polygon::SOURCE_CATEGORY_SEGMENT_END_POINT:
+        assert(cell.source_index() - points.size() < segments.size());
+        return segments[cell.source_index() - points.size()].from();
+        break;
+    default:
+        assert(false && "getSourcePoint should only be called on point cells!\n");
+        break;
+    }
+}
+
+PolygonsPointIndex VoronoiUtils::getSourcePointIndex(const vd_t::cell_type& cell, const std::vector<Point>& points, const std::vector<Segment>& segments)
+{
+    assert(cell.contains_point());
+    assert(cell.source_category() != boost::polygon::SOURCE_CATEGORY_SINGLE_POINT);
+    switch (cell.source_category())
+    {
+    case boost::polygon::SOURCE_CATEGORY_SEGMENT_START_POINT:
+    {
+        assert(cell.source_index() - points.size() < segments.size());
+        PolygonsPointIndex ret = segments[cell.source_index() - points.size()];
+        ++ret;
+        return ret;
+        break;
+    }
+    case boost::polygon::SOURCE_CATEGORY_SEGMENT_END_POINT:
+    {
+        assert(cell.source_index() - points.size() < segments.size());
+        return segments[cell.source_index() - points.size()];
+        break;
+    }
+    default:
+        assert(false && "getSourcePoint should only be called on point cells!\n");
+        break;
+    }
+}
+
+const VoronoiUtils::Segment& VoronoiUtils::getSourceSegment(const vd_t::cell_type& cell, const std::vector<Point>& points, const std::vector<Segment>& segments)
+{
+    assert(cell.contains_segment());
+    return segments[cell.source_index() - points.size()];
+}
+
+
+std::vector<Point> VoronoiUtils::discretizeParabola(const Point& p, const Segment& segment, Point s, Point e, coord_t approximate_step_size, float transitioning_angle)
+{
+    std::vector<Point> discretized;
+    // x is distance of point projected on the segment ab
+    // xx is point projected on the segment ab
+    Point a = segment.from();
+    Point b = segment.to();
+    Point ab = b - a;
+    Point as = s - a;
+    Point ae = e - a;
+    coord_t ab_size = vSize(ab);
+    coord_t sx = dot(as, ab) / ab_size;
+    coord_t ex = dot(ae, ab) / ab_size;
+    coord_t sxex = ex - sx;
+    
+    Point ap = p - a;
+    coord_t px = dot(ap, ab) / ab_size;
+    
+    Point pxx = LinearAlg2D::getClosestOnLine(p, a, b);
+    Point ppxx = pxx - p;
+    coord_t d = vSize(ppxx);
+    PointMatrix rot = PointMatrix(turn90CCW(ppxx));
+    
+    if (d == 0)
+    {
+        discretized.emplace_back(s);
+        discretized.emplace_back(e);
+        return discretized;
+    }
+    
+    float marking_bound = atan(transitioning_angle * 0.5);
+    coord_t msx = - marking_bound * d; // projected marking_start
+    coord_t mex = marking_bound * d; // projected marking_end
+    coord_t marking_start_end_h = msx * msx / (2 * d) + d / 2;
+    Point marking_start = rot.unapply(Point(msx, marking_start_end_h)) + pxx;
+    Point marking_end = rot.unapply(Point(mex, marking_start_end_h)) + pxx;
+    coord_t dir = 1;
```

Comment:
Better to use `int` here. The result is pretty much the same, but the type is better documenting. See my previous comment here: https://github.com/Ultimaker/CuraEngine/pull/1210#discussion_r391085244

```suggestion
    int dir = 1;
```